onnxruntime/src/tensor/ort_tensor.rs

```diff
@@ -37,38 +38,103 @@ where
 {
     pub(crate) fn from_array<'m>(
         memory_info: &'m MemoryInfo,
+        allocator_ptr: *mut sys::OrtAllocator,
         mut array: Array<T, D>,
     ) -> Result<OrtTensor<'t, T, D>>
     where
         'm: 't, // 'm outlives 't
     {
+        // where onnxruntime will write the tensor data to
         let mut tensor_ptr: *mut sys::OrtValue = std::ptr::null_mut();
         let tensor_ptr_ptr: *mut *mut sys::OrtValue = &mut tensor_ptr;
-        let tensor_values_ptr: *mut std::ffi::c_void = array.as_mut_ptr() as *mut std::ffi::c_void;
-        assert_ne!(tensor_values_ptr, std::ptr::null_mut());
 
         let shape: Vec<i64> = array.shape().iter().map(|d: &usize| *d as i64).collect();
         let shape_ptr: *const i64 = shape.as_ptr();
         let shape_len = array.shape().len() as u64;
 
-        let status = unsafe {
-            g_ort().CreateTensorWithDataAsOrtValue.unwrap()(
-                memory_info.ptr,
-                tensor_values_ptr,
-                (array.len() * std::mem::size_of::<T>()) as u64,
-                shape_ptr,
-                shape_len,
-                T::tensor_element_data_type(),
-                tensor_ptr_ptr,
-            )
-        };
-        status_to_result(status).map_err(OrtError::CreateTensorWithData)?;
-        assert_ne!(tensor_ptr, std::ptr::null_mut());
+        match T::tensor_element_data_type() {
+            TensorElementDataType::Float
+            | TensorElementDataType::Uint8
+            | TensorElementDataType::Int8
+            | TensorElementDataType::Uint16
+            | TensorElementDataType::Int16
+            | TensorElementDataType::Int32
+            | TensorElementDataType::Int64
+            | TensorElementDataType::Double
+            | TensorElementDataType::Uint32
+            | TensorElementDataType::Uint64 => {
```

Comment:
I would have put the `TensorElementDataType::String` first, and the remaining with a fall-through `_` to simplify this.

Reply:
Unless I'm missing something, I hesitate to use a `_` here since the remaining types like f16 or Complex don't seem like they would Just Work™ like the existing primitives. Thoughts?

Reply:
Or, put differently, the _ bit me for string _outputs_ because the compiler didn't yell at me to consider whether or not the existing logic would work for the newly added enum variant, and this seems like an opportunity for the same issue.